homeassistant/config.py

```diff
@@ -827,21 +826,87 @@ async def merge_packages_config(
     return config
 
 
-async def async_process_component_config(  # noqa: C901
-    hass: HomeAssistant, config: ConfigType, integration: Integration
+@overload
+async def async_process_component_config(
+    hass: HomeAssistant,
+    config: ConfigType,
+    integration: Integration,
+) -> ConfigType:
+    ...
+
+
+@overload
+async def async_process_component_config(
+    hass: HomeAssistant,
+    config: ConfigType,
+    integration: Integration,
+    *,
+    raise_on_failure: Literal[True],
+) -> ConfigType:
+    ...
+
+
+@overload
+async def async_process_component_config(
+    hass: HomeAssistant,
+    config: ConfigType,
+    integration: Integration,
+    *,
+    raise_on_failure: Literal[False] | bool,
+) -> ConfigType | None:
+    ...
+
+
+async def async_process_component_config(
+    hass: HomeAssistant,
+    config: ConfigType,
+    integration: Integration,
+    *,
+    raise_on_failure: bool = True,
 ) -> ConfigType | None:
     """Check component configuration and return processed configuration.
 
-    Returns None on error.
+    This method must be run in the event loop.
+    """
+    if raise_on_failure:
+        return await _async_process_component_config(hass, config, integration, True)
+    try:
+        return await _async_process_component_config(hass, config, integration, False)
+    except HomeAssistantError as ex:
+        _LOGGER.exception(ex)
+        return None
+
+
+async def _async_process_component_config(  # noqa: C901
+    hass: HomeAssistant,
+    config: ConfigType,
+    integration: Integration,
+    raise_on_platform_setup_errors: bool,
+) -> ConfigType:
+    """Check component configuration and return processed configuration.
 
     This method must be run in the event loop.
     """
     domain = integration.domain
+
+    def _raise_and_notify(ex: Exception, domain: str, config: ConfigType) -> None:
+        """Notify and raise an exception for a failing config."""
+        async_notify_setup_error(hass, domain, None)
+        message, _, config_file, line = _format_config_error(ex, domain, config, None)
+        placeholders = {
+            "domain": domain,
+            "error": str(ex),
```

Comment:
This means we don't use the humanized version of the error message here.
I would suggest to, in this PR, just use the message and in a follow-up PR add translations with placeholders for line numbers etc.

Reply:
We do use a humanized version of the error for reloads. The caller needs to catch the `ConfigValidationError` and re-raise it to a `ServiceValidationError`, it can make use of the `translation_placeholders`.

Reply:
https://github.com/home-assistant/core/pull/102410/files#diff-b327426a8b69bb17cc2626a2cc65cdf97da0f0edfd3670a52fb00cab899c274dR172-R190